File: demo/multiclass_classification/train.py
#! /usr/bin/python
import sys
import numpy as np
sys.path.append('../../wrapper/')
import xgboost as xgb

# label need to be 0 to num_class -1
data = np.loadtxt('./dermatology.data', delimiter=',',converters={33: lambda x:int(x == '?'), 34: lambda x:int(x)-1 } )
sz = data.shape

train = data[:int(sz[0] * 0.7), :]
test = data[int(sz[0] * 0.7):, :]

train_X = train[:,0:33]
train_Y = train[:, 34]


test_X = test[:,0:33]
test_Y = test[:, 34]

xg_train = xgb.DMatrix( train_X, label=train_Y)
xg_test = xgb.DMatrix(test_X, label=test_Y)
# setup parameters for xgboost
param = {}
# use softmax multi-class classification
param['objective'] = 'multi:softmax'
# scale weight of positive examples
param['eta'] = 0.1
param['max_depth'] = 6
param['silent'] = 1
param['nthread'] = 4
param['num_class'] = 6

watchlist = [ (xg_train,'train'), (xg_test, 'test') ]
num_round = 5
bst = xgb.train(param, xg_train, num_round, watchlist );
# get prediction
pred = bst.predict( xg_test );

print ('predicting, classification error=%f' % (sum( int(pred[i]) != test_Y[i] for i in range(len(test_Y))) / float(len(test_Y)) ))

# do the same thing again, but output probabilities
param['objective'] = 'multi:softprob'
bst = xgb.train(param, xg_train, num_round, watchlist );
# Note: this convention has been changed since xgboost-unity
# get prediction, this is in 1D array, need reshape to (ndata, nclass)
yprob = bst.predict( xg_test ).reshape( test_Y.shape[0], 6 )
ylabel = np.argmax(yprob, axis=1)

print ('predicting, classification error=%f' % (sum( int(ylabel[i]) != test_Y[i] for i in range(len(test_Y))) / float(len(test_Y)) ))

File: demo/multiclass_classification/train.R
library(data.table)
library(xgboost)

if (!file.exists("./dermatology.data")) {
  download.file(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/dermatology/dermatology.data",
    "dermatology.data",
    method = "curl"
  )
}

df <- fread("dermatology.data", sep = ",", header = FALSE)

df[, `:=`(V34 = as.integer(ifelse(V34 == "?", 0L, V34)),
          V35 = V35 - 1L)]

idx <- sample(nrow(df), size = round(0.7 * nrow(df)), replace = FALSE)

train <- df[idx,]
test <- df[-idx,]

train_x <- train[, 1:34]
train_y <- train[, V35]

test_x <- test[, 1:34]
test_y <- test[, V35]

xg_train <- xgb.DMatrix(data = as.matrix(train_x), label = train_y)
xg_test = xgb.DMatrix(as.matrix(test_x), label = test_y)

params <- list(
  objective = 'multi:softmax',
  num_class = 6,
  max_depth = 6,
  nthread = 4,
  eta = 0.1
)

watchlist = list(train = xg_train, test = xg_test)

bst <- xgb.train(
  params = params,
  data = xg_train,
  watchlist = watchlist,
  nrounds = 5
)

pred <- predict(bst, xg_test)
error_rate <- sum(pred != test_y) / length(test_y)
print(paste("Test error using softmax =", error_rate))

# do the same thing again, but output probabilities
params$objective <- 'multi:softprob'
bst <- xgb.train(params, xg_train, nrounds = 5, watchlist)

pred_prob <- predict(bst, xg_test)

pred_mat <- matrix(pred_prob, ncol = 6, byrow = TRUE)
# validation
# rowSums(pred_mat)

pred_label <- apply(pred_mat, 1, which.max) - 1L
error_rate = sum(pred_label != test_y) / length(test_y)
print(paste("Test error using softprob =", error_rate))
